### ISSUE TITLE
Bug: [no-unsafe-argument] doesn't flag unsafe argument with a type constraint

### ISSUE THREAD

**Issue by ronami** [typescript-eslint/typescript-eslint#10415]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3hkmCMAXPABQZEoCuAtgEYgwCU8AvAHzwBuSAlsANwAoUJFgJk6LDkwAmIgB4AKvBAAPHCmBp4lWvVakiCpm049%2BAgQHpL8AGZRuELchhwwGCAE8BUjHmJ48FBaUCieDIJWNmgAFkgUEMB2DhA%2BuBjSAUEhYRECQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolp9oAc1gBbRC3RRE0aB2iRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare const test1: (t: number) => void;
declare const test2: <T extends number>(t: T) => void;

// fails correctly
test1(1 as any);

// should fail
test2(1 as any);
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unsafe-argument": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect all use-cases to be consistent and report this as an error.

### Actual Result

The use-cases with a type constraint doesn't report this as an error.

### Additional Info

This seems similar to #10314, and it looks like the fix should be similar.

I'll add that I see similar inconsistencies on `no-unsafe-assignment` and `no-unsafe-return` (false positives, in which the rule reports unnecessarily):

```ts
// no-unsafe-return
declare const test1: (callback: () => unknown) => void;
declare const test2: <T>(callback: () => T) => void;
declare const test3: <T extends unknown>(callback: () => T) => void;

// passes correctly
test1(() => 1 as any);

// should pass
test2(() => 1 as any);

// should pass
test3(() => 1 as any);

// no-unsafe-assignment
declare const test4: (t: { t: unknown }) => void;
declare const test5: <T extends unknown>(t: { t: T }) => void;

// pass correctly
test4({ t: 1 as any });

// should pass
test5({ t: 1 as any });
```

I'll be happy also to report them (separately or as part of this issue) if this makes sense.

**Comment by kirkwaiblinger** [comment]
I'm starting to think that the root cause might be more akin to https://github.com/typescript-eslint/typescript-eslint/issues/6951, https://github.com/typescript-eslint/typescript-eslint/issues/9007 (and all the duplicates mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/9007#issuecomment-2080161404) than to #10314.

**Comment by kirkwaiblinger** [comment]
that is to say, I think the rule is seeing it as an `any` -> `any` assignment, see the following examples:
```ts
// should flag, does
test2<number>(1 as any);

// should flag, doesn't
test2<any>(1 as any);
```

**Comment by ronami** [comment]
> that is to say, I think the rule is seeing it as an `any` -> `any` assignment, see the following examples:
> 
> ```ts
> // should flag, does
> test2<number>(1 as any);
> 
> // doesn't flag.... and maybe shouldn't (??)
> test2<any>(1 as any);
> ```

Yes! I think it's exactly like @bradzacher explain here: https://github.com/typescript-eslint/typescript-eslint/issues/6951#issuecomment-1520945881.

**Comment by ronami** [comment]
> that is to say, I think the rule is seeing it as an `any` -> `any` assignment, see the following examples:
> 
> ```ts
> // should flag, does
> test2<number>(1 as any);
> 
> // doesn't flag.... and maybe shouldn't (??)
> test2<any>(1 as any);
> ```

Yes! I think it's very similar to what @bradzacher explained here: https://github.com/typescript-eslint/typescript-eslint/issues/6951#issuecomment-1520945881.
